File: EVANS/frmMantUsuarios.vb
Imports System.Data.SqlClient

Public Class frmMantUsuarios

    Dim objUsuario As New clsUsuario
    Dim bolAccion As Boolean = True

    Private Sub frmMantCliente_FormClosing(ByVal sender As Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles Me.FormClosing

        Try
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If

            LimpiarRAM()

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub frmMantCliente_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        Try
            Me.Top = 0
            Me.Left = 0

            TabControl1.TabPages(1).Enabled = False
            txtBuscar.Enabled = False

            btnGrabar.Enabled = False
            btnCancelar.Enabled = False
            btnEditar.Enabled = False
            btnBuscar.Enabled = False

            optBuscar.Checked = True

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub optTodos_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles optTodos.CheckedChanged

        Try
            txtBuscar.Clear()
            lvListado.Items.Clear()
            Dim lstLista As New List(Of clsUsuario)

            lstLista = objUsuario.Listar()

            For i As Integer = 0 To lstLista.Count - 1
                Dim itemLista As ListViewItem
                itemLista = lvListado.Items.Add(lstLista(i).Codigo)
                itemLista.SubItems.Add(lstLista(i).NombreCompleto)
                itemLista.SubItems.Add(lstLista(i).NombreUsuario)
            Next

            txtBuscar.Enabled = False
            btnBuscar.Enabled = False

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub optBuscar_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles optBuscar.CheckedChanged

        Try
            lvListado.Items.Clear()
            txtBuscar.Enabled = True
            txtBuscar.Focus()
            btnBuscar.Enabled = True

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub lvListado_DoubleClick(ByVal sender As Object, ByVal e As System.EventArgs) Handles lvListado.DoubleClick

        Try
            Dim strCodigo As String
            strCodigo = lvListado.SelectedItems.Item(0).SubItems(0).Text

            objUsuario.Codigo = strCodigo
            If objUsuario.BuscarXCodigo() Then

                TabControl1.TabPages(1).Enabled = True
                TabControl1.TabPages(0).Enabled = False
                TabControl1.SelectTab(1)
                LimpiarControles(TabControl1, 1)

                txtCodigo.Text = objUsuario.Codigo
                txtEmpleado.Text = objUsuario.NombreCompleto
                txtUsuario.Text = objUsuario.NombreUsuario
                txtClave.Text = objUsuario.Clave
                txtRepetir.Text = objUsuario.Clave
                cbEstado.Text = objUsuario.Estado.Descripcion
                chkAdmin.Checked = IIf(objUsuario.Administrador = 1, True, False)

                'Desactivar controles
                DesactivarControles(TabControl1, 1, True)

                'Comportamiento de botones
                btnNuevo.Enabled = True
                btnGrabar.Enabled = False
                btnEditar.Enabled = True
                btnCancelar.Enabled = True

                btnGrabar.Text = "Grabar"

            Else
                Return
            End If

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub btnNuevo_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnNuevo.Click

        Try
            objUsuario = New clsUsuario

            'Comportamiento del control TabControl
            TabControl1.TabPages(1).Enabled = True
            TabControl1.TabPages(0).Enabled = False
            TabControl1.SelectTab(1)
            txtEmpleado.Focus()

            'Limpiar y activa cajas de texto
            LimpiarControles(TabControl1, 1)
            DesactivarControles(TabControl1, 1, False)
            txtCodigo.ReadOnly = True

            'Comportamiento de botones
            btnNuevo.Enabled = False
            btnGrabar.Enabled = True
            btnCancelar.Enabled = True

            objUsuario = New clsUsuario

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub btnGrabar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnGrabar.Click

        Try
            If txtEmpleado.Text = "" Or txtUsuario.Text = "" Or txtClave.Text = "" Or txtRepetir.Text = "" Or cbEstado.Text = "" Then
                MessageBox.Show("Datos incompletos", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Exit Sub
            End If

            If txtClave.Text = txtRepetir.Text Then
                objUsuario.Clave = txtClave.Text
            Else
                MessageBox.Show("La nueva clave no concuerda con la confirmación.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Exit Sub
            End If

            objUsuario.NombreCompleto = txtEmpleado.Text
            objUsuario.NombreUsuario = txtUsuario.Text
            objUsuario.Estado.Codigo = BuscarCodigo("esta_codigo", "esta_descripcion", cbEstado.Text, "Estado")
            objUsuario.Administrador = IIf(chkAdmin.Checked = True, 1, 0)

            If bolAccion = True Then
                If objUsuario.Grabar() = False Then
                    Return
                End If
            Else
                If objUsuarioActual.Codigo = objUsuario.Codigo Or objUsuarioActual.Administrador = 1 Then
                    If objUsuario.Actualizar() = False Then
                        Return
                    Else
                        bolAccion = True
                    End If
                Else
                    MessageBox.Show("Solo un usuario administrador o el propietario de la cuenta pueden modificar este registro.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                    Return
                End If
                
            End If

            'Comportamiento del TabControl
            TabControl1.TabPages(0).Enabled = True
            TabControl1.TabPages(1).Enabled = False
            TabControl1.SelectTab(0)

            'Comportamiento de Botones
            btnNuevo.Enabled = True
            btnGrabar.Enabled = False
            btnEditar.Enabled = False
            btnCancelar.Enabled = False

            btnGrabar.Text = "Grabar"

            optBuscar.Checked = True

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub btnEditar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnEditar.Click

        Try
            'Activa las cajas de texto
            DesactivarControles(TabControl1, 1, False)
            txtCodigo.ReadOnly = True
            txtEmpleado.Focus()

            'Comportamiento de botones
            btnNuevo.Enabled = False
            btnGrabar.Enabled = True
            btnEditar.Enabled = False
            btnCancelar.Enabled = True

            btnGrabar.Text = "Actualizar"
            bolAccion = False

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub btnCancelar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancelar.Click

        Try
            'Comportamiento del TabControl
            TabControl1.TabPages(0).Enabled = True
            TabControl1.TabPages(1).Enabled = False
            TabControl1.SelectTab(0)

            bolAccion = True
            btnGrabar.Text = "Grabar"

            'Comportamiento de botones
            btnNuevo.Enabled = True
            btnGrabar.Enabled = False
            btnEditar.Enabled = False
            btnCancelar.Enabled = False

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub btnSalir_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSalir.Click

        Try
            Me.Close()

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub btnBuscar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnBuscar.Click

        Try
            If txtBuscar.Text = "" Then
                lvListado.Items.Clear()
                MessageBox.Show("Ingrese nombre a buscar", "Datos insuficientes", MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtBuscar.Focus()
            Else
                lvListado.Items.Clear()
                objUsuario.NombreCompleto = txtBuscar.Text

                Dim drLista As SqlDataReader

                drLista = objUsuario.BuscarXNombre()

                While drLista.Read()
                    Dim itemLista As ListViewItem
                    itemLista = lvListado.Items.Add(NullToString(drLista(0)))
                    itemLista.SubItems.Add(NullToString(drLista(1)))
                    itemLista.SubItems.Add(NullToString(drLista(2)))
                    itemLista.SubItems.Add(NullToString(drLista(3)))
                End While
                drLista.Close()
            End If

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub txtBuscar_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles txtBuscar.KeyPress

        Try
            If e.KeyChar = ChrW(Keys.Enter) Then
                e.Handled = True
                btnBuscar.PerformClick()
            End If

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub


    Private Sub cbEstado_DropDown(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cbEstado.DropDown

        Try
            LLenarCombo("esta_codigo", "esta_descripcion", "Estado", cbEstado, False)
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub chkAdmin_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles chkAdmin.CheckedChanged

        If chkAdmin.Checked = True Then
            If objUsuarioActual.Administrador <> 1 Then
                MessageBox.Show("Solo un usuario administrador puede modificar esta opción.", "Mensaje", MessageBoxButtons.OK, MessageBoxIcon.Information)
                chkAdmin.Checked = False
            End If
        End If

    End Sub
End Class

File: src/EVANS.UI.WinForms/Identidad/frmMantUsuarios.cs
using EVANS.Application.Catalogo.DTOs;
using EVANS.Application.Catalogo.Queries;
using EVANS.Application.Common;
using EVANS.Application.Identidad.Commands;
using EVANS.Application.Identidad.DTOs;
using EVANS.Application.Identidad.Ports;
using EVANS.Application.Identidad.Queries;
using MediatR;

namespace EVANS.UI.WinForms.Identidad;

public partial class frmMantUsuarios : Form
{
    private readonly IMediator? _mediator;
    private readonly ICurrentSession? _currentSession;
    private IReadOnlyList<UsuarioCuentaDto> _usuarios = [];
    private IReadOnlyList<EstadoDto> _estados = [];
    private UsuarioCuentaDto? _selectedUsuario;
    private bool _isCreateMode = true;
    private bool _suppressAdminCheck;

    public frmMantUsuarios() => InitializeComponent();

    public frmMantUsuarios(IMediator mediator, ICurrentSession currentSession)
    {
        _mediator = mediator;
        _currentSession = currentSession;
        InitializeComponent();
        WireEvents();
        ApplyInitialState();
    }

    internal int SelectedTabIndex => tabControl1.SelectedIndex;
    internal bool DetailsTabEnabled => tabPageDetalles.Enabled;
    internal bool ListingTabEnabled => tabPageListado.Enabled;
    internal bool BuscarEnabled => btnBuscar.Enabled;
    internal bool BuscarTextEnabled => txtBuscar.Enabled;
    internal bool NuevoEnabled => btnNuevo.Enabled;
    internal bool GrabarEnabled => btnGrabar.Enabled;
    internal bool EditarEnabled => btnEditar.Enabled;
    internal bool CancelarEnabled => btnCancelar.Enabled;
    internal string GrabarText => btnGrabar.Text;
    internal string CodigoText => txtCodigo.Text;
    internal string EmpleadoText => txtEmpleado.Text;
    internal string UsuarioText => txtUsuario.Text;
    internal string StatusMessage => lblmsg.Text;
    internal int ListCount => lvListado.Items.Count;
    internal string? FirstListUsuario => lvListado.Items.Count == 0 ? null : lvListado.Items[0].SubItems[2].Text;
    internal bool AdminChecked => chkAdmin.Checked;

    internal void SetTestSearchText(string value) => txtBuscar.Text = value;
    internal void SetTestEmpleado(string value) => txtEmpleado.Text = value;
    internal void SetTestUsuario(string value) => txtUsuario.Text = value;
    internal void SetTestClave(string value) => txtClave.Text = value;
    internal void SetTestRepetir(string value) => txtRepetir.Text = value;
    internal void SetTestEstado(int codigo)
    {
        EnsureEstadoItem(codigo, codigo == 1 ? "ACTIVO" : codigo.ToString());
        cbEstado.SelectedValue = codigo;
    }
    internal void SetTestAdmin(bool value) => chkAdmin.Checked = value;
    internal void BeginNewForTest() => btnNuevo_Click();
    internal void BeginEditForTest() => btnEditar_Click();

    internal async Task OpenFirstForTestAsync()
    {
        if (lvListado.Items.Count == 0)
            return;

        await OpenUsuarioAsync(int.Parse(lvListado.Items[0].SubItems[0].Text));
    }

    internal async Task LoadUsuariosAsync()
    {
        _usuarios = await Mediator.Send(new BuscarUsuariosQuery(null));
        BindUsuarios(_usuarios);
    }

    internal async Task<bool> BuscarAsync(bool showMessages = false)
    {
        if (string.IsNullOrWhiteSpace(txtBuscar.Text))
        {
            lvListado.Items.Clear();
            return ShowInformation("Ingrese nombre a buscar", "Datos insuficientes", showMessages);
        }

        _usuarios = await Mediator.Send(new BuscarUsuariosQuery(txtBuscar.Text.Trim()));
        BindUsuarios(_usuarios);
        SetStatus(DefaultStatusMessage);
        return true;
    }

    internal async Task<bool> SaveAsync(bool showMessages = false)
    {
        if (string.IsNullOrWhiteSpace(txtEmpleado.Text) ||
            string.IsNullOrWhiteSpace(txtUsuario.Text) ||
            string.IsNullOrWhiteSpace(txtClave.Text) ||
            string.IsNullOrWhiteSpace(txtRepetir.Text) ||
            cbEstado.SelectedValue is null)
        {
            return ShowError("Datos incompletos", showMessages);
        }

        if (!string.Equals(txtClave.Text, txtRepetir.Text, StringComparison.Ordinal))
            return ShowError("La nueva clave no concuerda con la confirmación.", showMessages);

        if (!_isCreateMode && !CanEditSelectedUsuario())
            return ShowError("Solo un usuario administrador o el propietario de la cuenta pueden modificar este registro.", showMessages);

        var estadoCodigo = (int)cbEstado.SelectedValue;
        if (_isCreateMode)
        {
            var result = await Mediator.Send(new CrearUsuarioCommand(
                txtUsuario.Text,
                txtClave.Text,
                txtEmpleado.Text,
                chkAdmin.Checked,
                estadoCodigo));
            if (!result.IsSuccess)
                return ShowFailure(result, showMessages);
        }
        else
        {
            var result = await Mediator.Send(new ActualizarUsuarioCommand(
                int.Parse(txtCodigo.Text),
                txtUsuario.Text,
                txtClave.Text,
                txtEmpleado.Text,
                chkAdmin.Checked,
                estadoCodigo));
            if (!result.IsSuccess)
                return ShowFailure(result, showMessages);
        }

        await ReturnToListingAsync();
        return true;
    }

    private void WireEvents()
    {
        optTodos.CheckedChanged += async (_, _) => await optTodos_CheckedChanged();
        optBuscar.CheckedChanged += (_, _) => optBuscar_CheckedChanged();
        lvListado.DoubleClick += async (_, _) => await lvListado_DoubleClick();
        btnNuevo.Click += (_, _) => btnNuevo_Click();
        btnGrabar.Click += async (_, _) => await SaveAsync(showMessages: true);
        btnEditar.Click += (_, _) => btnEditar_Click();
        btnCancelar.Click += async (_, _) => await ReturnToListingAsync();
        btnSalir.Click += (_, _) => Close();
        btnBuscar.Click += async (_, _) => await BuscarAsync(showMessages: true);
        txtBuscar.KeyPress += async (_, e) =>
        {
            if (e.KeyChar == (char)Keys.Enter)
            {
                e.Handled = true;
                await BuscarAsync(showMessages: true);
            }
        };
        cbEstado.DropDown += async (_, _) => await LoadEstadosAsync();
        chkAdmin.CheckedChanged += (_, _) => chkAdmin_CheckedChanged();
    }

    private void ApplyInitialState()
    {
        Top = 0;
        Left = 0;

        tabPageDetalles.Enabled = false;
        txtBuscar.Enabled = false;

        btnGrabar.Enabled = false;
        btnCancelar.Enabled = false;
        btnEditar.Enabled = false;
        btnBuscar.Enabled = false;

        optBuscar.Checked = true;
    }

    private async Task optTodos_CheckedChanged()
    {
        if (!optTodos.Checked)
            return;

        txtBuscar.Clear();
        _usuarios = await Mediator.Send(new BuscarUsuariosQuery(null));
        BindUsuarios(_usuarios);
        txtBuscar.Enabled = false;
        btnBuscar.Enabled = false;
    }

    private void optBuscar_CheckedChanged()
    {
        if (!optBuscar.Checked)
            return;

        lvListado.Items.Clear();
        txtBuscar.Enabled = true;
        txtBuscar.Focus();
        btnBuscar.Enabled = true;
    }

    private async Task lvListado_DoubleClick()
    {
        if (lvListado.SelectedItems.Count == 0)
            return;

        await OpenUsuarioAsync(int.Parse(lvListado.SelectedItems[0].SubItems[0].Text));
    }

    private async Task OpenUsuarioAsync(int codigo)
    {
        var usuario = await Mediator.Send(new ObtenerUsuarioPorCodigoQuery(codigo));
        if (usuario is null)
            return;

        _selectedUsuario = usuario;
        await LoadEstadosAsync();
        EnsureEstadoItem(usuario.EstadoCodigo, usuario.EstadoCodigo.ToString());

        tabPageDetalles.Enabled = true;
        tabPageListado.Enabled = false;
        tabControl1.SelectTab(tabPageDetalles);

        txtCodigo.Text = usuario.Codigo.ToString();
        txtEmpleado.Text = usuario.NombreCompleto;
        txtUsuario.Text = usuario.NombreUsuario;
        txtClave.Text = usuario.Clave;
        txtRepetir.Text = usuario.Clave;
        cbEstado.SelectedValue = usuario.EstadoCodigo;
        SetAdminChecked(usuario.EsAdministrador);
        SetDetailInputsEnabled(false);

        btnNuevo.Enabled = true;
        btnGrabar.Enabled = false;
        btnEditar.Enabled = true;
        btnCancelar.Enabled = true;
        btnGrabar.Text = "Grabar";
    }

    private void btnNuevo_Click()
    {
        _isCreateMode = true;
        _selectedUsuario = null;
        tabPageDetalles.Enabled = true;
        tabPageListado.Enabled = false;
        tabControl1.SelectTab(tabPageDetalles);

        ClearDetails();
        SetDetailInputsEnabled(true);
        txtCodigo.ReadOnly = true;
        txtEmpleado.Focus();

        btnNuevo.Enabled = false;
        btnGrabar.Enabled = true;
        btnCancelar.Enabled = true;
        btnEditar.Enabled = false;
        btnGrabar.Text = "Grabar";
    }

    private void btnEditar_Click()
    {
        SetDetailInputsEnabled(true);
        txtCodigo.ReadOnly = true;
        txtEmpleado.Focus();

        btnNuevo.Enabled = false;
        btnGrabar.Enabled = true;
        btnEditar.Enabled = false;
        btnCancelar.Enabled = true;
        btnGrabar.Text = "Actualizar";
        _isCreateMode = false;
    }

    private async Task ReturnToListingAsync()
    {
        tabPageListado.Enabled = true;
        tabPageDetalles.Enabled = false;
        tabControl1.SelectTab(tabPageListado);

        _isCreateMode = true;
        _selectedUsuario = null;
        btnGrabar.Text = "Grabar";
        btnNuevo.Enabled = true;
        btnGrabar.Enabled = false;
        btnEditar.Enabled = false;
        btnCancelar.Enabled = false;
        optBuscar.Checked = true;
        lvListado.Items.Clear();
        await Task.CompletedTask;
    }

    private async Task LoadEstadosAsync()
    {
        if (_estados.Count != 0)
            return;

        _estados = await Mediator.Send(new ListEstadosQuery());
        BindEstados();
    }

    private void BindEstados()
    {
        cbEstado.DisplayMember = nameof(EstadoDto.Descripcion);
        cbEstado.ValueMember = nameof(EstadoDto.Codigo);
        cbEstado.DataSource = _estados.ToList();
    }

    private void EnsureEstadoItem(int codigo, string descripcion)
    {
        if (_estados.Any(estado => estado.Codigo == codigo))
        {
            BindEstados();
            return;
        }

        _estados = _estados.Append(new EstadoDto(codigo, descripcion)).ToList();
        BindEstados();
    }

    private void SetDetailInputsEnabled(bool enabled)
    {
        txtCodigo.Enabled = enabled;
        txtEmpleado.Enabled = enabled;
        txtUsuario.Enabled = enabled;
        txtClave.Enabled = enabled;
        txtRepetir.Enabled = enabled;
        cbEstado.Enabled = enabled;
        chkAdmin.Enabled = enabled;
    }

    private void ClearDetails()
    {
        txtCodigo.Clear();
        txtEmpleado.Clear();
        txtUsuario.Clear();
        txtClave.Clear();
        txtRepetir.Clear();
        cbEstado.SelectedIndex = -1;
        SetAdminChecked(false);
    }

    private void BindUsuarios(IEnumerable<UsuarioCuentaDto> usuarios)
    {
        lvListado.Items.Clear();
        foreach (var usuario in usuarios)
        {
            var item = lvListado.Items.Add(usuario.Codigo.ToString());
            item.SubItems.Add(usuario.NombreCompleto);
            item.SubItems.Add(usuario.NombreUsuario);
        }
    }

    private void chkAdmin_CheckedChanged()
    {
        if (_suppressAdminCheck || !chkAdmin.Checked || CurrentUserIsAdmin)
            return;

        ShowInformation("Solo un usuario administrador puede modificar esta opción.", "Mensaje", showMessages: false);
        SetAdminChecked(false);
    }

    private void SetAdminChecked(bool value)
    {
        _suppressAdminCheck = true;
        chkAdmin.Checked = value;
        _suppressAdminCheck = false;
    }

    private bool CanEditSelectedUsuario()
    {
        if (CurrentUserIsAdmin)
            return true;

        var currentUserName = _currentSession?.Current?.Usuario.NombreUsuario;
        var selectedUserName = _selectedUsuario?.NombreUsuario;
        return !string.IsNullOrWhiteSpace(currentUserName) &&
               string.Equals(currentUserName, selectedUserName, StringComparison.OrdinalIgnoreCase);
    }

    private bool CurrentUserIsAdmin => _currentSession?.Current?.Usuario.EsAdministrador == true;

    private bool ShowFailure<T>(Result<T> result, bool showMessages) =>
        ShowError(result.Error ?? "No se pudo guardar el usuario.", showMessages);

    private bool ShowError(string message, bool showMessages)
    {
        SetStatus(message);
        if (showMessages)
            MessageBox.Show(message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
        return false;
    }

    private bool ShowInformation(string message, string caption, bool showMessages)
    {
        SetStatus(message);
        if (showMessages)
            MessageBox.Show(message, caption, MessageBoxButtons.OK, MessageBoxIcon.Information);
        txtBuscar.Focus();
        return false;
    }

    private void SetStatus(string message) => lblmsg.Text = message;

    private const string DefaultStatusMessage = "Los campos marcados con asterisco (*) son obligatorios";

    private IMediator Mediator => _mediator
        ?? throw new InvalidOperationException("Use the IMediator constructor at runtime. The parameterless constructor is for the WinForms designer only.");
}
